lua/nvim_lsp/angularls.lua

```diff
@@ -0,0 +1,69 @@
+local configs = require 'nvim_lsp/configs'
+local util = require 'nvim_lsp/util'
+
+local server_name = 'angularls'
+local bin_name = server_name
+local install_loc = util.base_install_dir .. '/' .. server_name
+local script_loc = install_loc .. '/node_modules/@angular/language-server/index.js'
+local bin_loc = install_loc .. '/node_modules/.bin/angularls'
+
+local installer = util.npm_installer {
+  server_name = server_name,
+  packages = { '@angular/language-server' },
+  binaries = { bin_name },
+  -- angular-language-service doesn't expose a binary, so we create an execution wrapper.
+  post_install_script =
+    'echo "#! /bin/sh\n' .. 'node ' .. script_loc .. ' \\$*' .. '" > ' .. bin_loc .. '\n' ..
+    'chmod +x ' .. bin_loc
+}
+
+local function get_probe_dir()
+  local project_root = util.find_node_modules_ancestor(vim.fn.getcwd())
+
+  return project_root and (project_root .. '/node_modules') or ''
+end
+
+local default_probe_dir = get_probe_dir()
+
+configs[server_name] = {
+  default_config = {
+    cmd = {
+      bin_loc,
+      '--stdio',
+      '--tsProbeLocations', default_probe_dir,
+      '--ngProbeLocations', default_probe_dir
+    },
+    filetypes = {'typescript', 'html', 'typescriptreact', 'typescript.tsx'},
+    root_dir = util.root_pattern('angular.json', 'tsconfig.json', 'package.json', '.git'),
+    on_new_config = function(new_config)
+      local install_info = installer.info()
+      local new_probe_dir = get_probe_dir()
```

Comment:
@h-michael  Do you know if there is a way to get the new resolved root directory in the `on_new_config` hook? I think using the `vim.fn.getcwd()` isn't quite the right way here.

Reply:
I think in this case(node project), probe directory equal the directory where `node_modules` exists.
So I think that that is `util.root_pattern('package.json').."/node_modules"`.

Reply:
Actually this doesn't work because we need to know the resolved root directory that caused this hook to execute. For example, I have two angular projects and I open index.html in one and index.html in the other. The probe directories always point to the initial cwd. 

I updated this and added an argument to the `on_new_config` hook that passes the newly resolved root directory. This seemed like the only way to get that newly resolved root to generate the new command and doesn't break any existing functionality.

Reply:
@steelsojka
Could you give me a little more detail?
Guess your example, is it a project with multiple node_modules in a mono repo?

Reply:
In an angular specific mono repo you only have one node_modules folder but multiple `package.json` files.

You can ignore my comment above. The issue I was running into is solved by providing the new root dir (from the `root_dir` function) to the `on_new_config` hook.

Reply:
I don't know why we should add root_dir argument.
Does it mean that we need to add "root_dir" argument because the root_dir may be different depending on the files opened in the same project?

Reply:
The problem is we need the config to generate the new command. You open a file from two separate projects in the same  vim instance. We get a command for each project because the roots are different. We need to set the probe directories based off of that new root dir.

Now the problem with the current implementation is we don't have access to that new root dir. You can't use `new_config.root_dir` because it is set AFTER the hooks are invoked. So the solution is you either pass it as an argument to the hook OR you assign `root_dir` to `new_config` prior to calling the `on_new_config` callback. I'm fine with either solution, but we need some way to access the new root directory.

Reply:
> You open a file from two separate projects in the same vim instance. We get a command for each project because the roots are different. We need to set the probe directories based off of that new root dir.

Our built-in language server client can't you multiple same language server in same Nvim instance.
`lsp.foo.setup()` function is set one-shot autocmd to start a language server.

Reply:
Looking at the code, it looks like it can. Basically if the `root_dir` function returns a new root directory that hasn't had a client created for that root, it starts a new client. I think that is one of the purposes of the `root_dir` function and the `server_per_root_dir_manager`.

https://github.com/neovim/nvim-lsp/blob/master/lua/nvim_lsp/util.lua#L246

So when a new config is generated, we need to have that new root directory to generate the new command for the new server that is being spawned.

Reply:
@h-michael Is the point above make sense? Is there anything else required for this to go in?